File: redo/osxaltest/spinbox.m
// 31 july 2015
#import "osxaltest.h"

// leave a whole lot of space around the alignment rect, just to be safe
@interface tSpinboxContainer : NSView
@end

@implementation tSpinboxContainer

- (NSEdgeInsets)alignmentRectInsets
{
	return NSEdgeInsetsMake(50, 50, 50, 50);
}

@end

int nspinbox = 0;

@implementation tSpinbox {
	tSpinboxContainer *c;
	NSTextField *t;
	NSStepper *s;
	id<tControl> parent;
	NSLayoutPriority horzpri, vertpri;
}

- (id)init
{
	self = [super init];
	if (self) {
		NSMutableDictionary *views;
		NSArray *constraints;

		self->c = [[tSpinboxContainer alloc] initWithFrame:NSZeroRect];
		[self->c setTranslatesAutoresizingMaskIntoConstraints:NO];

		self->t = [[NSTextField alloc] initWithFrame:NSZeroRect];
[self->t setStringValue:[NSString stringWithFormat:@"%d", nspinbox]];
nspinbox++;
		[self->t setSelectable:YES];
		[self->t setFont:[NSFont systemFontOfSize:[NSFont systemFontSizeForControlSize:NSRegularControlSize]]];
		[self->t setBordered:NO];
		[self->t setBezelStyle:NSTextFieldSquareBezel];
		[self->t setBezeled:YES];
		[[self->t cell] setLineBreakMode:NSLineBreakByClipping];
		[[self->t cell] setScrollable:YES];
		[self->t setTranslatesAutoresizingMaskIntoConstraints:NO];
		[self->c addSubview:self->t];

		self->s = [[NSStepper alloc] initWithFrame:NSZeroRect];
		[self->s setIncrement:1];
		[self->s setValueWraps:NO];
		[self->s setAutorepeat:YES];
		[self->s setTranslatesAutoresizingMaskIntoConstraints:NO];
		[self->c addSubview:self->s];

		views = [NSMutableDictionary new];
		[views setObject:self->t forKey:@"t"];
		[views setObject:self->s forKey:@"s"];

		constraints = [NSLayoutConstraint constraintsWithVisualFormat:@"H:|[t]-[s]|" options:0 metrics:nil views:views];
		[self->c addConstraints:constraints];
		constraints = [NSLayoutConstraint constraintsWithVisualFormat:@"V:|[t]|" options:0 metrics:nil views:views];
		[self->c addConstraints:constraints];
		constraints = [NSLayoutConstraint constraintsWithVisualFormat:@"V:|[s]|" options:0 metrics:nil views:views];
		[self->c addConstraints:constraints];

		[views release];

		self->parent = nil;

		self->horzpri = [self->c contentHuggingPriorityForOrientation:NSLayoutConstraintOrientationHorizontal];
		self->vertpri = [self->c contentHuggingPriorityForOrientation:NSLayoutConstraintOrientationVertical];
	}
	return self;
}

- (void)tSetParent:(id<tControl>)p addToView:(NSView *)v relayout:(BOOL)relayout
{
	self->parent = p;
	[v addSubview:self->c];
	if (relayout)
		[self tRelayout];
}

- (void)tFillAutoLayout:(tAutoLayoutParams *)p
{
	// reset the hugging priority
	[self->c setContentHuggingPriority:self->horzpri forOrientation:NSLayoutConstraintOrientationHorizontal];
	[self->c setContentHuggingPriority:self->vertpri forOrientation:NSLayoutConstraintOrientationVertical];
	p->view = self->c;
	p->attachLeft = YES;
	p->attachTop = YES;
	p->attachRight = YES;
	p->attachBottom = YES;
	p->nonStretchyWidthPredicate = @"(==96)";		// TODO on the text field only
}

- (void)tRelayout
{
	if (self->parent != nil)
		[self->parent tRelayout];
}

@end

File: redo/osxaltest/spinbox.swift
// 31 july 2015
import Cocoa

var nspinbox = 0

class Spinbox : NSView, Control {
	private var t: NSTextField
	private var s: NSStepper
	private var parent: Control?

	init() {
		var cell: NSTextFieldCell

		self.t = NSTextField(frame: NSZeroRect)
self.t.stringValue = "\(nspinbox)"
nspinbox++
		self.t.selectable = true
		self.t.font = NSFont.systemFontOfSize(NSFont.systemFontSizeForControlSize(NSControlSize.RegularControlSize))
		self.t.bordered = false
		self.t.bezelStyle = NSTextFieldBezelStyle.SquareBezel
		self.t.bezeled = true
		cell = self.t.cell() as! NSTextFieldCell
		cell.lineBreakMode = NSLineBreakMode.ByClipping
		cell.scrollable = true
		self.t.translatesAutoresizingMaskIntoConstraints = false

		// make the textbox grow horizontally and vertically
		setHorzHuggingPri(self.t, myNSLayoutPriorityDefaultLow)
		setVertHuggingPri(self.t, myNSLayoutPriorityDefaultLow)

		self.s = NSStepper(frame: NSZeroRect)
		self.s.increment = 1
		self.s.valueWraps = false
		self.s.autorepeat = true
		self.s.translatesAutoresizingMaskIntoConstraints = false

		// make the spinner grow vertically but not horizontally
		// TODO Required instead?
		setHorzHuggingPri(self.s, myNSLayoutPriorityDefaultHigh)
		setVertHuggingPri(self.s, myNSLayoutPriorityDefaultLow)

		self.parent = nil

		super.init(frame: NSZeroRect)
		self.translatesAutoresizingMaskIntoConstraints = false

		self.addSubview(self.t)
		self.addSubview(self.s)
		var views = [
			"t":	self.t,
			"s":	self.s,
		]
		var constraints = mkconstraints("H:|[t]-[s]|", views)
		self.addConstraints(constraints)
		constraints = mkconstraints("V:|[t]|", views)
		self.addConstraints(constraints)
		constraints = mkconstraints("V:|[s]|", views)
		self.addConstraints(constraints)
	}

	required init?(coder: NSCoder) {
		fatalError("can't use this constructor, sorry")
	}

	// TODO leave only the required amount of space around the alignment rect
	// TODO even with this the stepper sometimes gets cut off at the bottom *anyway*
	override var alignmentRectInsets: NSEdgeInsets {
		get {
			return NSEdgeInsetsMake(50, 50, 50, 50)
		}
	}

	func View() -> NSView {
		return self
	}

	func SetParent(p: Control) {
		self.parent = p
	}

	// TODO justify this
	// TODO no really, is this height the right way to go?
	// TODO restrict width to the text field only?
	override var intrinsicContentSize: NSSize {
		get {
			var s = super.intrinsicContentSize
			s.width = 96
			s.height = max(self.t.intrinsicContentSize.height, self.s.intrinsicContentSize.height)
			return s
		}
	}
}
